### PR TITLE
Avoid broken `strip` versions on 32-bit Linux

### LINKED ISSUE #2339: Linker problem when installing alex on CentOS 5 with newer cabal-install
Okay, yikes, esoteric issue title, I know.

On a 32-bit CentOS 5 machine, I've got three installations of `cabal-install` and `Cabal`:

```
$ /opt/ghc-7.6.3/bin/cabal --version
cabal-install version 1.18.0.3
using version 1.18.1.3 of the Cabal library
$ /opt/ghc-7.8.3/bin/cabal --version
cabal-install version 1.20.0.3
using version 1.20.0.1 of the Cabal library
$ cabal --version
cabal-install version 1.22.0.0
using version 1.22.0.0 of the Cabal library
```

Keeping GHC versions the same at 7.8.3, just switching `cabal`s, I try installing `alex` into a fresh sandbox. The 1.18.0.3 install goes swimmingly. The 1.20.0.3 and 1.22.0.0 installs however fail to link the `alex` executable at the end of the build:

```
...
Linking dist/dist-sandbox-79015d48/build/alex/alex ...
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x3bb): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x543): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x883): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xb9f): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xd5b): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xee3): more undefined references to `Threefish_256_Process_Block' follow
collect2: ld returned 1 exit status
Failed to install alex-3.1.4
Updating world file...
cabal: Error: some packages failed to install:
alex-3.1.4 failed during the building phase. The exception was:
ExitFailure 1
```

I notice that the final invocation of `ghc` to build the executable has more linker flags in the newer versions. Maybe something is causing the linker to miss the symbol from `tf-random`'s cbits?

I've posted full logs here: https://gist.github.com/acfoltzer/c46274098956d03e6f79

It's worth noting that this does not occur on a similarly-configured 64-bit machine, nor any of our newer CentOS machines.

Thanks for any ideas you might have.

### PR DESCRIPTION
**Pull request by acfoltzer**
Fixes #2339 

Tested on Yosemite, Windows 7, CentOS 5 32/64-bit, CentOS 6 32/64-bit

The only thing I don't feel 100% about here is how we parse the numeric version out of the `strip --version` output. It's pretty inconsistent between platforms and versions and how the binutils were built, so we try to parse the first numeric word of the output and pull out a major and minor version for comparison. At worst this will prevent library stripping from working on 32-bit Linux systems with non-standard `strip --version` output.

### CODE REVIEW

**Comment by tibbe** [comment]
Inconsistent flags indeed:

```
$ strip --version
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: unrecognized option: --version
Usage: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip [-AnuSXx] [-] [-d filename] [-s filename] [-R filename] [-o output] file [...]
```

**Comment by acfoltzer** [comment]
Yeah, non-GNU `strip` doesn't seem to have any way of reporting a version. Since the version info isn't used except on 32-bit Linux though, it should be fine.

File: Cabal/Distribution/Simple/Program/Strip.hs [hunk 1]
```diff
@@ -53,6 +56,17 @@ stripLib verbosity (Platform _arch os) conf path = do
                -- libraries with lots identically named modules. See
                -- #1784.
                return()
+    Linux | arch == I386 ->
+      -- Versions of 'strip' on 32-bit Linux older than 2.18 are
+      -- broken. See #2339.
+      let (Just okVersion) = simpleParse ">= 2.18"
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Strip.hs]: This looks a bit fishy, why not use `orLaterVersion`?

Comment (acfoltzer) [inline Cabal/Distribution/Simple/Program/Strip.hs]: Is there another way to make `Version`s other than parsing? I thought `">= 2.18"` would parse into `orLaterVersion <Version 2.18>`; this seemed like the only way to get such a `VersionRange`.

Comment (23Skidoo) [inline Cabal/Distribution/Simple/Program/Strip.hs]: `orLaterVersion (Version [2,18] [])`?

Comment (acfoltzer) [inline Cabal/Distribution/Simple/Program/Strip.hs]: D'oh, of course. I thought cabal used a special opaque `Version`. Patch incoming

**Pull request merged**